### PR TITLE
Port MYFACES-4469

### LINKED ISSUE #25111: MYFACES-4469: IllegalArgumentException occurs in occurs in FacesConfigurator.purgeConfiguration
**Describe the bug**  
If an application uses a custom factory via META-INF/services (such as javax.faces.application.ApplicationFactory), then a purge failure can occur **when in development mode** when an update occurs to the faces-config.xml. Additionally, the server must not restart the application (which can occur).  

```
java.lang.IllegalArgumentException: object is not an instance of declaring class
…
at java.lang.reflect.Method.invoke(Method.java:508)
org.apache.myfaces.config.FacesConfigurator.purgeConfiguration(FacesConfigurator.java:555)
at org.apache.myfaces.config.FacesConfigurator.update(FacesConfigurator.java:485)
at org.apache.myfaces.lifecycle.LifecycleImpl.execute(LifecycleImpl.java:136)
```

**Steps to Reproduce**  
This one is difficult to reproduce since if an update to the faces-config.xml occurs in Liberty, then the application will updated/restarted entirely within the server and not enter the problematic method.  However, some faces developers have reported this occurring when an application is built in one timezone, but then deployed in another.  

Occurs when in the project stage is set to Development, workaround is to use Production mode instead.

**Expected behavior**  
No exceptions should occur. 

**Diagnostic information:**  
 - OpenLiberty Version: ALL
 - Affected feature(s) jsf-2.2, jsf-2.3, faces-3.0, and faces-4.0
 - Java Version: N/A
 - server.xml configuration : N/A

**Additional context**  
https://issues.apache.org/jira/projects/MYFACES/issues/MYFACES-4469
The faces-4.0 feature was fixed via the 4.0.1 update under #25354

### PR DESCRIPTION
**Pull request by volosied**
fixes #25111

No tests created as I can't reproduce this problem easily. Manual testing was performed via breakpoints in Eclipse (where I modified the faces-config timestamp in order to go into the _purgeConfiguration_ method).

### CODE REVIEW

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

File: dev/com.ibm.ws.jsf.2.2/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 1]
```diff
@@ -463,50 +464,21 @@ else if (!dirPath.startsWith("/META-INF") && dirPath.endsWith("/"))
...
+
     public void update()
     {
-        //Google App Engine does not allow to get last modified time of a file; 
-        //and when an application is running on GAE there is no way to update faces config xml file.
-        //thus, no need to check if the config file is modified.
-        if (ContainerUtils.isRunningOnGoogleAppEngine(_externalContext))
-        {
-            return;
-        }
```
Comment (volosied) [inline dev/com.ibm.ws.jsf.2.2/src/org/apache/myfaces/config/FacesConfigurator.java]: Add back in

File: dev/com.ibm.ws.org.apache.myfaces.2.3/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 2]
```diff
@@ -465,48 +466,18 @@ else if (!dirPath.startsWith("/META-INF") && dirPath.endsWith("/"))
 
     public void update()
     {
-        //Google App Engine does not allow to get last modified time of a file; 
-        //and when an application is running on GAE there is no way to update faces config xml file.
-        //thus, no need to check if the config file is modified.
-        if (ContainerUtils.isRunningOnGoogleAppEngine(_externalContext))
-        {
-            return;
-        }
```
Comment (volosied) [inline dev/com.ibm.ws.org.apache.myfaces.2.3/src/org/apache/myfaces/config/FacesConfigurator.java]: add back in

File: dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/application/ApplicationFactoryImpl.java [hunk 3]
```diff
@@ -26,12 +26,14 @@
 import jakarta.faces.context.ExternalContext;
 import jakarta.faces.context.FacesContext;
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/application/ApplicationFactoryImpl.java]: new line

File: dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 4]
```diff
@@ -453,48 +454,18 @@ else if (!dirPath.startsWith("/META-INF") && dirPath.endsWith("/"))
 
     public void update()
     {
-        //Google App Engine does not allow to get last modified time of a file; 
-        //and when an application is running on GAE there is no way to update faces config xml file.
-        //thus, no need to check if the config file is modified.
-        if (ContainerUtils.isRunningOnGoogleAppEngine(_externalContext))
-        {
-            return;
-        }
```
Comment (volosied) [inline dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java]: add back in

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by pnicolucci** [comment]
@volosied  can you point me to the commits in the MyFaces project so I can validate the changes here?

**Comment by volosied** [comment]
Original Commit:  https://github.com/apache/myfaces/commit/fb5d71e9a563b3be7da8ff8c8bf272f425c610bf 

Follow up Fix:  https://github.com/apache/myfaces/commit/f842c82c9c688fdf804fa8f488723ccdf5551e26


You cad find my PRs here too: https://github.com/apache/myfaces/pulls?q=is%3Apr+is%3Aclosed+4469

File: dev/io.openliberty.org.apache.myfaces.4.0/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 5]
```diff
@@ -147,6 +147,7 @@
 import org.apache.myfaces.spi.ResourceLibraryContractsProviderFactory;
 import org.apache.myfaces.util.ExternalSpecifications;
 import org.apache.myfaces.util.NavigationUtils;
+import org.apache.myfaces.util.Purgeable;
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.4.0/src/org/apache/myfaces/config/FacesConfigurator.java]: import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

Both of these were removed in the upstream changes.

File: dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 6]
```diff
@@ -503,100 +481,61 @@ public void update()
...
-    private boolean purgeConfiguration() throws NoSuchMethodException, IllegalAccessException, InvocationTargetException
+    private void purgeConfiguration()
     {
-
-        Method appFactoryPurgeMethod;
-        Method renderKitPurgeMethod;
-        Method lifecyclePurgeMethod;
-        Method facesContextPurgeMethod;
-
-        System.out.println("PURGE");
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java]: I didn't see this upstream so I'm assuming you just had this here for initial testing locally?

File: dev/io.openliberty.org.apache.myfaces.4.0/src/org/apache/myfaces/config/FacesConfigurator.java (hunk 5 above)
Comment (pnicolucci) [reply dev/io.openliberty.org.apache.myfaces.4.0/src/org/apache/myfaces/config/FacesConfigurator.java]: This applies across all versions you updated.

File: dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java [hunk 7]
```diff
@@ -503,100 +481,61 @@ public void update()
...
+            if (instance instanceof FacesWrapper)
             {
-                return purgeMethod;
+                instance = ((FacesWrapper<T>) instance).getWrapped();
             }
-            if (instance instanceof FacesWrapper)
+            else
             {
-                instance = ((FacesWrapper)instance).getWrapped();
+                return;
```
Comment (pnicolucci) [inline dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java:534]: Looking at the upstream commit it looks like this was missed and needs to be applied to your 3.0.x PR

Comment (pnicolucci) [reply dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java:534]: https://github.com/volosied/myfaces/blob/9ccc1a48bf7a0ba9c46725a1b1708e2992f72957/impl/src/main/java/org/apache/myfaces/config/FacesConfigurator.java#L531

Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java:534]: Fixed: https://github.com/apache/myfaces/commit/218dd53ce0d38263fb74f6d49b70b5b15671eeeb

File: dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java (hunk 6 above)
Comment (volosied) [reply dev/io.openliberty.org.apache.myfaces.3.0/src/org/apache/myfaces/config/FacesConfigurator.java]: Oops. Yes, it was for testing

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jsf.2.3_fat,com.ibm.ws.jsf.2.2_fat,io.openliberty.org.apache.myfaces.4.0_fat

**Comment by volosied** [comment]
#libby

**Pull request merged**